### ISSUE TITLE
ROCR Runtime Build Fails with GCC 8.x

### ISSUE THREAD

**Issue by sjug** [RadeonOpenCompute/ROCR-Runtime#44]
Build log:
```
-- The C compiler identification is GNU 8.2.1
-- The CXX compiler identification is GNU 8.2.1
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found LibElf: /usr/lib/libelf.so
-- Performing Test ELF_GETSHDRSTRNDX
-- Performing Test ELF_GETSHDRSTRNDX - Success
-- Looking for __NR_memfd_create
-- Looking for __NR_memfd_create - found
-- Configuring done
-- Generating done
-- Build files have been written to: /tmp/makepkg/hsa-rocr/src/ROCR-Runtime/build
Scanning dependencies of target hsa-link
Scanning dependencies of target hsa-runtime64.so-link
Scanning dependencies of target hsa-runtime64
[  0%] Built target hsa-link
[  0%] Built target hsa-runtime64.so-link
[  2%] Building CXX object CMakeFiles/hsa-runtime64.dir/core/runtime/amd_blit_sdma.cpp.o
[  5%] Building CXX object CMakeFiles/hsa-runtime64.dir/core/util/small_heap.cpp.o
[ 11%] Building CXX object CMakeFiles/hsa-runtime64.dir/core/util/timer.cpp.o
[ 11%] Building CXX object CMakeFiles/hsa-runtime64.dir/core/util/lnx/os_linux.cpp.o
[ 16%] Building CXX object CMakeFiles/hsa-runtime64.dir/core/runtime/amd_blit_kernel.cpp.o
[ 16%] Building CXX object CMakeFiles/hsa-runtime64.dir/core/runtime/default_signal.cpp.o
[ 19%] Building CXX object CMakeFiles/hsa-runtime64.dir/core/runtime/amd_gpu_agent.cpp.o
[ 22%] Building CXX object CMakeFiles/hsa-runtime64.dir/core/runtime/amd_cpu_agent.cpp.o
[ 25%] Building CXX object CMakeFiles/hsa-runtime64.dir/core/runtime/hsa_ven_amd_loader.cpp.o
[ 27%] Building CXX object CMakeFiles/hsa-runtime64.dir/core/runtime/amd_aql_queue.cpp.o
[ 30%] Building CXX object CMakeFiles/hsa-runtime64.dir/core/runtime/amd_loader_context.cpp.o
[ 33%] Building CXX object CMakeFiles/hsa-runtime64.dir/core/runtime/amd_memory_region.cpp.o
[ 36%] Building CXX object CMakeFiles/hsa-runtime64.dir/core/runtime/amd_topology.cpp.o
[ 44%] Building CXX object CMakeFiles/hsa-runtime64.dir/core/runtime/hsa_api_trace.cpp.o
[ 44%] Building CXX object CMakeFiles/hsa-runtime64.dir/core/runtime/host_queue.cpp.o
[ 50%] Building CXX object CMakeFiles/hsa-runtime64.dir/core/runtime/interrupt_signal.cpp.o
[ 50%] Building CXX object CMakeFiles/hsa-runtime64.dir/core/runtime/intercept_queue.cpp.o
[ 44%] Building CXX object CMakeFiles/hsa-runtime64.dir/core/runtime/hsa.cpp.o
[ 55%] Building CXX object CMakeFiles/hsa-runtime64.dir/core/runtime/hsa_ext_amd.cpp.o
[ 55%] Building CXX object CMakeFiles/hsa-runtime64.dir/core/runtime/hsa_ext_interface.cpp.o
[ 58%] Building CXX object CMakeFiles/hsa-runtime64.dir/core/runtime/ipc_signal.cpp.o
[ 61%] Building CXX object CMakeFiles/hsa-runtime64.dir/core/runtime/isa.cpp.o
[ 63%] Building CXX object CMakeFiles/hsa-runtime64.dir/core/runtime/runtime.cpp.o
[ 66%] Building CXX object CMakeFiles/hsa-runtime64.dir/core/runtime/cache.cpp.o
[ 69%] Building CXX object CMakeFiles/hsa-runtime64.dir/core/runtime/signal.cpp.o
[ 72%] Building CXX object CMakeFiles/hsa-runtime64.dir/core/runtime/queue.cpp.o
[ 77%] Building CXX object CMakeFiles/hsa-runtime64.dir/loader/executable.cpp.o
[ 77%] Building CXX object CMakeFiles/hsa-runtime64.dir/core/common/shared.cpp.o
[ 80%] Building CXX object CMakeFiles/hsa-runtime64.dir/core/common/hsa_table_interface.cpp.o
[ 83%] Building CXX object CMakeFiles/hsa-runtime64.dir/loader/loaders.cpp.o
[ 86%] Building CXX object CMakeFiles/hsa-runtime64.dir/libamdhsacode/amd_elf_image.cpp.o
[ 88%] Building CXX object CMakeFiles/hsa-runtime64.dir/libamdhsacode/amd_hsa_code_util.cpp.o
[ 91%] Building CXX object CMakeFiles/hsa-runtime64.dir/libamdhsacode/amd_hsa_locks.cpp.o
[ 94%] Building CXX object CMakeFiles/hsa-runtime64.dir/libamdhsacode/amd_options.cpp.o
[ 97%] Building CXX object CMakeFiles/hsa-runtime64.dir/libamdhsacode/amd_hsa_code.cpp.o
/tmp/makepkg/hsa-rocr/src/ROCR-Runtime/src/core/runtime/runtime.cpp: In member function ‘void core::Runtime::SetLinkCount(size_t)’:
/tmp/makepkg/hsa-rocr/src/ROCR-Runtime/src/core/runtime/runtime.cpp:260:71: error: ‘void* memset(void*, int, size_t)’ clearing an object of non-trivial type ‘__gnu_cxx::__alloc_traits<std::allocator<core::Runtime::LinkInfo>, core::Runtime::LinkInfo>::value_type’ {aka ‘struct core::Runtime::LinkInfo’}; use assignment or value-initialization instead [-Werror=class-memaccess]
          link_matrix_.size() * sizeof(hsa_amd_memory_pool_link_info_t));
                                                                       ^
In file included from /tmp/makepkg/hsa-rocr/src/ROCR-Runtime/src/core/runtime/runtime.cpp:43:
/tmp/makepkg/hsa-rocr/src/ROCR-Runtime/src/core/inc/runtime.h:97:10: note: ‘__gnu_cxx::__alloc_traits<std::allocator<core::Runtime::LinkInfo>, core::Runtime::LinkInfo>::value_type’ {aka ‘struct core::Runtime::LinkInfo’} declared here
   struct LinkInfo {
          ^~~~~~~~
cc1plus: all warnings being treated as errors
make[2]: *** [CMakeFiles/hsa-runtime64.dir/build.make:349: CMakeFiles/hsa-runtime64.dir/core/runtime/runtime.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:73: CMakeFiles/hsa-runtime64.dir/all] Error 2
make: *** [Makefile:152: all] Error 2
==> ERROR: A failure occurred in build().
    Aborting...
```

[This diff](https://github.com/oleid/ROCR-Runtime/commit/b68be8f5101f00b583797781606bf906de973fff.diff) fixes GCC 8.x compatibility:
```
diff --git a/src/core/runtime/runtime.cpp b/src/core/runtime/runtime.cpp
index 0625c6f..4875c2b 100644
--- a/src/core/runtime/runtime.cpp
+++ b/src/core/runtime/runtime.cpp
@@ -256,8 +256,7 @@ void Runtime::SetLinkCount(size_t num_link) {
   const size_t last_index = GetIndexLinkInfo(0, num_link);
   link_matrix_.resize(last_index);
 
-  memset(&link_matrix_[0], 0,
-         link_matrix_.size() * sizeof(hsa_amd_memory_pool_link_info_t));
+  std::fill(link_matrix_.begin(), link_matrix_.end(), LinkInfo());
 }
 
 void Runtime::RegisterLinkInfo(uint32_t node_id_from, uint32_t node_id_to,
```

**Comment by jlgreathouse** [comment]
Dupe of #42 

That said, we've applied a very similar patch internally. We're still working with our program managers for how exactly to "release" this patch. We don't really want to update roc-1.9.x publicly, because then someone who wanted to build exactly what's in our .deb/.rpm files would be building against something different. I'll update this ticket when we've figured out the right way to make this patch visible while not breaking any of our AMD release processes.
